user/test-super/com/google/gwt/dev/jjs/super/com/google/gwt/dev/jjs/test/Java17Test.java

```diff
@@ -76,4 +78,143 @@ private void checkIfCompiled(Shape square, Shape circle) {
     assertTrue(square instanceof Square);
     assertTrue(circle instanceof Circle);
   }
+
+  /**
+   * Sample nested record. Has no components. Implements a single interface
+   */
+  record InnerRecord() implements Comparable<InnerRecord> {
+    /**
+     * Simple accessor-looking method to ensure it doesn't participate in hashcode/tostring.
+     */
+    public String value() {
+      return "Hello";
+    }
+
+    @Override
+    public int compareTo(InnerRecord other) {
+      return 0;
+    }
+  }
+  public void testRecordClasses() {
+    /**
+     * Sample local record.
+     */
+    record LocalRecord() {
+      @Override
+      public String toString() {
+        return "Example";
+      }
+    }
+
+    assertTrue(new InnerRecord() instanceof Record);
+    assertTrue(new InnerRecord() instanceof Comparable);
+
+    assertFalse(new InnerRecord().toString().contains("Hello"));
+    assertTrue(new InnerRecord().toString().startsWith("InnerRecord"));
+    assertEquals(0, new InnerRecord().hashCode());
+    assertEquals(new InnerRecord(), new InnerRecord());
+
+    assertEquals("Example", new LocalRecord().toString());
+
+    TopLevelRecord withValues = new TopLevelRecord("Banana", 7);
+    assertTrue(withValues.toString().contains("7"));
+    assertTrue(withValues.toString().contains("Banana"));
+    assertEquals("Banana", withValues.name());
+    assertEquals(7, withValues.count());
+  }
```

Comment:
For to string shouldnt we test for the full default format that is expected to match the JVM records to string and make sure the name and the value of each field present?

Reply:
My interpretation of the spec is that as long as it contains the name of the class, the name of each component, and the value of each component, we're within the spec. I did try to match pretty exactly to what OpenJDK will return.

https://docs.oracle.com/javase/specs/jls/se17/html/jls-8.html#jls-8.10
> A method public final String toString() that returns a string derived from the name of the record class and the names and string representations of every record component of R.
> The string representation of a record component c of an instance a of a record class is as follows:
> -   If the type of the record component c is a reference type, then the string representation is determined as if by invoking the toString method on the value of the component field c of a.
> -   If the type of the record component c is a primitive type T, then the string representation is determined as if by subjecting the value of the component field c of a to boxing conversion ([§5.1.7](https://docs.oracle.com/javase/specs/jls/se17/html/jls-5.html#jls-5.1.7)) and then invoking the method toString method of the wrapper class corresponding to T on the resulting object.


For the sake of respective the obfuscated type settings of a given application, I interpreted "the name of the record class" as "the value returned from `this.getClass().getSimpleName()`, so that the name will be consistently obfuscated, see ImplementRecordComponents.java:
```java
  private void implementToString(JRecordType type, JMethod method, SourceInfo info) {
    List<JExpression> args = new ArrayList<>();

    // Concatenate type with []s and component values, assigning to toStrExpr as we append
    // more concat operations. Using getClass().getSimpleName() rather than a string literal
    // allows the toString implementation to emit the obfuscated class name.
    JMethodCall getClass = new JMethodCall(info, new JThisRef(info, type), getClassMethod);
    JExpression toStrExpr = new JMethodCall(info, getClass, getSimpleNameMethod);
```